### PR TITLE
[FSK-75]: Add CreateFamily Mutation

### PR DESCRIPTION
**Pull request by mhmda-83**
add createFamily mutation and related tests.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -15,6 +16,8 @@ async function bootstrap() {
 
   const { port } = configService.get<ServerConfig>('server');
 
+  app.useGlobalPipes(new ValidationPipe());
```
Comment (pourdaavar) [inline src/main.ts]: To have no extra work for setup e2e testing, It's better to move this one to its own module or related module,
You can create RequestModule and set this validation to it, also in the future also any module that depends on to request operation would be placed here.

File: src/family/family.service.ts [hunk 2]
```diff
@@ -18,4 +22,40 @@ export class FamilyService {
...
+    const familySlug = slug(data.name);
+    const extendedData = {
+      ...data,
+      slug: familySlug,
+    };
+
+    // eslint-disable-next-line fp/no-let
+    let family: Family | null = null;
+    // eslint-disable-next-line fp/no-let
+    let counter = 0;
```
Comment (pourdaavar) [inline src/family/family.service.ts]: ignore linter rules Repeatedly :)

File: src/family/family.service.ts [hunk 3]
```diff
@@ -18,4 +22,40 @@ export class FamilyService {
...
+        });
+      } catch (error: any) {
+        if (error?.code === 'P2002' && error?.meta?.target?.includes('slug')) {
+          counter++;
+        } else {
+          this.logger.error({ errorOnCreateFamily: error });
+          throw error;
+        }
+      }
+    }
```
Comment (pourdaavar) [inline src/family/family.service.ts]: As mentioned before, it's not good practice to have to `await` in the loop,
this solution is not appropriate,

The solution I can represent right now:
 1. make a find query to check proposed slug ( like a proposed slug ) is assigned before or not
 2. then based on the result decide to change `slug` to another one.

In this way, there is only 2 separate query, but no query in the loop.

File: src/family/family.service.spec.ts [hunk 4]
```diff
@@ -57,4 +60,33 @@ describe('FamilyService', () => {
...
+      const data: CreateFamilyInput = { name: faker.name.fullName() };
+      const method = jest
+        .spyOn(prisma.family, 'create')
+        .mockResolvedValue(familyStub);
+      await service.create(data);
+
+      expect(method).toHaveBeenCalledWith({
+        data: { ...data, slug: slug(data.name) },
+      });
+    });
```
Comment (pourdaavar) [inline src/family/family.service.spec.ts:80]: this test case belongs to family service, not repo!
As I understood you write a test case to ensure `create` method on the family repo is called with the correct args or not,
you should test `create` method on the service called with appropriate args or not

File: src/family/family.service.spec.ts [hunk 5]
```diff
@@ -57,4 +60,33 @@ describe('FamilyService', () => {
...
+    it('should throw unrecognized error', async () => {
+      jest
+        .spyOn(prisma.family, 'create')
+        .mockRejectedValue(new Error('Wrong Input'));
+
+      await expect(
+        service.create({ name: faker.name.fullName() }),
+      ).rejects.toThrow();
+    });
+  });
```
Comment (pourdaavar) [inline src/family/family.service.spec.ts:100]: I think it's better to write a test case for the expected result too

**Review by pourdaavar — CHANGES_REQUESTED** [review]

File: src/family/family.service.spec.ts (hunk 5 above)
Comment (mhmda-83) [reply src/family/family.service.spec.ts:100]: Done

File: src/family/family.service.ts (hunk 3 above)
Comment (mhmda-83) [reply src/family/family.service.ts]: Done.

this way is not perfect but in our scale is totally perfect.

File: src/family/family.service.ts (hunk 2 above)
Comment (mhmda-83) [reply src/family/family.service.ts]: Fixed

File: src/family/family.service.spec.ts (hunk 4 above)
Comment (mhmda-83) [reply src/family/family.service.spec.ts:80]: I don't understand your point here.

File: src/family/family.service.ts (hunk 3 above)
Comment (pourdaavar) [reply src/family/family.service.ts]: At least it's better than before!

File: src/main.ts (hunk 1 above)
Comment (mhmda-83) [reply src/main.ts]: I didn't understand your point here.
what problem will produce if we using global validation pipe?

File: src/family/family.service.spec.ts (hunk 4 above)
Comment (pourdaavar) [reply src/family/family.service.spec.ts:80]: you should write test for `create` method on service is called with expected args, not create method on `prisma.family.create`

**Review by pourdaavar — CHANGES_REQUESTED** [review]

File: src/family/family.service.spec.ts (hunk 4 above)
Comment (mhmda-83) [reply src/family/family.service.spec.ts:80]: What's the point to check which parameters that I passed in service.create method? (I will call it directly in test case and there is no way to be wrong)

I should check is service.create method calls prisma service with correct parameters or not. (I called service.create on line 75)

File: src/main.ts (hunk 1 above)
Comment (pourdaavar) [reply src/main.ts]: in e2e setup, you should call ValidaitonPipe in each test file
in e2e you just bring AppModule but this validation is defined in `main.ts`.

Comment (mhmda-83) [reply src/main.ts]: I understood.
I will add ValidationPipe to AppModule provider.

https://stackoverflow.com/questions/59355841/how-to-apply-global-pipes-during-e2e-tests

Comment (mhmda-83) [reply src/main.ts]: Done

File: src/family/family.service.spec.ts (hunk 4 above)
Comment (pourdaavar) [reply src/family/family.service.spec.ts:80]: Actually, both of them are useless :)
no difference between them,
in the context of unit test, we try to ensure is our unit work correctly and whether We expect it or not,
based on this definition, checking if the Prisma family `create` method called with appropriate args is not belong to this definition

Comment (mhmda-83) [reply src/family/family.service.spec.ts:80]: Yes, you're right.

What's your suggestion?

Comment (pourdaavar) [reply src/family/family.service.spec.ts:80]: It's ok, let's close this. 
we should ask @ASafaeirad

**Review by pourdaavar — APPROVED** [review]

File: src/family/family.resolver.spec.ts [hunk 6]
```diff
@@ -136,4 +138,48 @@ describe('family Query', () => {
...
+      query,
+    });
+
+    const firstSlug = firstRequestResult.data?.['createFamily']?.slug;
+    const secondSlug = secondRequestResult.data?.['createFamily']?.slug;
+
+    expect(firstSlug).toBeTruthy();
+    expect(secondSlug).toBeTruthy();
+    expect(firstSlug).not.toEqual(secondSlug);
+
```
Comment (MohammadForoutan) [inline src/family/family.resolver.spec.ts:176]: I'm not sure we need to make sure family store in db and check it or not
but I do it in my ticket what is you opinion?
Is this  an overdo or not?

File: src/main.ts (hunk 1 above)
Comment (MohammadForoutan) [reply src/main.ts]: I think we can do it later

**Review by MohammadForoutan — CHANGES_REQUESTED** [review]

**Pull request merged**